### PR TITLE
Add KLL Sketch Functions

### PR DESCRIPTION
**Pull request by ZacBlanco**
## Description

This change adds KLL Sketch support from the Apache DataSketches library. One of the benefits of KLL sketches is that the implementation supports more datatypes than the existing quantile digest and T-Digest implementations. It also has provably-bounded error compared to the T-Digest

The following datatype is added

- `kllsketch(T)`: Basically a wrapper around `varbinary`, similar to `tdigest` and `qdigest` types

The following functions are added

- `sketch_kll`: computes a KLL sketch
- `sketch_kll_with_k`: computes a kll sketch with the given value for k
- `sketch_kll_quantile`: queries the sketch for a particular quantile
- `sketch_kll_rank`: queries the sketch for a particular rank (inverse of the quantile function)

## Motivation and Context

Eventual use by the optimizer after #21236. But also good to have since HMS is [releasing support for them soon](https://issues.apache.org/jira/browse/HIVE-26221).

## Impact

Two new aggregation functions, two new scalar functions

## Test Plan

Standard aggregation unit tests and scalar function unit tests

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== RELEASE NOTES ==

General Changes
* New support for Apache DataSketches KLL sketch with the sketch_kll and related family of functions
```

### CODE REVIEW

File: presto-docs/src/main/sphinx/functions/sketch.rst
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/sketch.rst:42]:
```suggestion
about the KLL sketch, see the Apache Datasketches `KLL Sketch documentation`_.
```

**Review by steveburnett — CHANGES_REQUESTED** [review]
Only a very tiny nit, otherwise LGTM!

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java
Comment (Yuhta) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:21]: There are some pitfalls (mainly memory-wise) we use KLL in presto_cpp, not sure if they are applicable to this Java implementation or not:
1. The sketch is initialized with a fairly large pre-allocated array depending on the error bound.  This would cause large memory usage when the number of groups are high.
2. The serialized form is not very compact and can occupy more memory than it needs.  This is ok for the partial aggregation side but would cause memory issue on the final aggregation side.

Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:21]: Thanks for bringing up these points! Was there anything you did to mitigate them?

re: 1. I haven't checked the code, but didn't consider the performance with a high number of groups since the main target use case for introducing the function is to provide table and partition-level statistics for the optimizer. If you have any thoughts on optimization I would happily discuss.

re: 2: I haven't measured the serialized size but according to the charts on the Apache Datasketches website, a "standard" k value of 200 produces sketches in the range of 2-4k bytes serialized with floats after consuming 1M+ items. Let's generously assume that with doubles, we would consume double the space, so about 8k.

kll size.

Do you know which parameters effect the number of partial aggregations sent to the final agg? I assumed it's the size of the cluster but, I could be wrong. Even if coalescing 1k KLL sketches in one node this should only consume 8k * 1000 = ~8MB of memory. I feel that shouldn't pose any issue. But it really depends on how many sketches are sent to the final agg step.

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:21]: @Yuhta what sort of memory issues were observed?

Comment (Yuhta) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:21]: For 1 it's easy to fix in the sketch itself, just need to make the sketch grows the bottom layer when the number of elements it keeps is small.  You may need to submit PR to `datasketches` project to fix it though, or copy the code over.

For 2 the size of serialization also depends on how the input data looks like.  The worst case comparing to Q/T-digest comes from a lot of identical inputs.  In this case Q-digest does not increase the memory usage but KLL does, because it keeps samples of inputs instead of distribution.  This multiplied by the number of groups and the fan-in factor of final aggregation sometimes blows it up.  One mitigation we did is to compact the sketch before the serialization, this does not resolve the problem completely but we have been good since then on production data.

Both issues can be fixed later though, I just mention it here so we are not surprised when these issues showing up.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:89]: The problem with this is it only estimates the size of the list, not the size of the memory used in aggregate.  It needs to be `SIZE_OF(KLLSketch) * sketches.sizeOf() + INSTANCE_SIZE`.

One way of doing this is to keep a running total in the grouped aggregation state, and increment/decrement it whenever sketches are added.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:61]: Is this the same thing as the size in memory?  If not, we need to use a more accurate method that captures the size in memory for accounting purposes.  If they're off, then such error can grow in magnitude for large group by or when the input is skewed.

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java
Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:61]: I always assumed it was the serialized size. Docs on the interface would be nice...

Anyways, there isn't going to be a perfect way to calculate the size, especially for variable-width types as we discussed offline. For fixed-width types, I think a heuristic might be okay. I replaced this with the following function:

```java
        static final long SKETCH_INSTANCE_SIZE = ClassLayout.parseClass(KllItemsSketch.class).instanceSize();
        @Override
        public long getEstimatedSize()
        {
            if (sketch != null) {
                // Estimating the memory is difficult without direct access to
                // the underlying implementation. We will attempt to give a
                // close estimate without underestimating to prevent crashes.
                // Overestimates might claim more resources, and prevent more
                // queries from running.
                // Since retained entries generally dominate the memory usage.
                return (SKETCH_INSTANCE_SIZE +
                        INSTANCE_SIZE +
                        (long) (sketch.getNumRetained() * retainedItemSize * 1.2));
            }

            return INSTANCE_SIZE;
        }
```

I ran some rough benchmarks using JOL's `GraphLayout#parseInstance` which attempts to accurately calculate retained memory of a single object. I compared it against the above java function to test the accuracy. Here are the results:

image

image

At around 2048 items (2^11), the error percentage drops under 1 percent, which decreases up to 0.1% after the sketch has seen > 100k items.
I might be able to modify the heuristic to account for sketches which have seen a smaller number of item. from the stream. But I think the error is within a reasonable margin? What are your thoughts on the reasonable margin of error?

Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:61]: FWIW I also experimented against the T-Digest using a compression factor giving comparable sizes to the above chart:
image

The maximum error here is roughly 20%, while the minimum is ~0.27%

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:61]: @ZacBlanco although this seems like a moderate risk, I am wary of doing something Presto-side since upgrades might be unstable and cause the heuristic to become out much more inaccurate.

Can you estimate how does the serialized size compare to JOL's estimation?  If they're close, then perhaps we could use that (plus a fudge factor), as it seems less likely that the data structure would change without a corresponding change to the serialized layout.  Additionally, you could update the grouped version to keep track one-by-one using the same heuristic.

Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:61]: Sorry for the delay on this one. I spent some time analyzing to come up with a decent heuristic that
we can use for the meantime based on the serialized size of the sketch.

While it's possible to make an addition to the sketch APIs to report utilized memory it's difficult
as they make heavy use of abstract classes which makes the calculation more complex than
most of the `ClassLayout#parseClass` instances in the Presto codebase. I started a small
implementation that might work for our case but proposing and implementing that API will probably
take some time to get accepted by the community.

While we wait for that change to land in the datasketches upstream, I ran a regression analysis to
determine the best adjustment factor to use. For each datatype, you can model the true in-memory
size, $t$ as $t(x) = f_1 x + f_2$ where $x$ is the size of the serialized sketch. I think this
actually models the true size quite well because when serialized, the size will be dominated by the
number of retained items. But those retained items take up less space serialized than when in the
JVM object always by a constant factor (e.g. an instance of `Double` takes up 24 bytes, but when
serialized just 8 bytes). Any other overhead in the sketch is constant (e.g. min/max, seen count, etc)
which is why we need the constant factor adjustment. This method should work for both fixed-width
and variable-sized datatypes since the serialized size actually contains the values which contribute
to memory consumption - unlike the approach with retained items I tried above.

Using this model I ran the regression to find the best parameters of $f_1$ and $f_2$ for each
datatype. Some charts are below. For strings and numeric types the regression function seems to
pretty closely track the true usage. For booleans there is a discrepancy where the regression over-
estimates by a bit.

image

These were the parameters I found:

| Type | $f_1$ | $f_2$ |
|-----|-----|----|
| Double |  929.0746716729514 | 3.3328879969131457 |
| Long | 1008.5483794752541 | 4.234749475485517 |
| String | 345.60095530779705 | 14.12193288555095
| Boolean | 929.920850240635 | 3.3325974295999283 |

I would probably encode them into a static hashmap that gets referenced when querying the estimated size.

The ideal solution is for sure to get something upstream, but if we really care about getting accurate sizes for grouped queries, then this will be a pretty good stopgap solution with minimal error until then. The accuracy issues with booleans I am not very concerned about because there isn't really a great use case for booleans and KLL sketches anyways.

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:61]: Thank you for working on this detailed analysis.  I agree that your heuristic seems to be a good stopgap for now.

Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java]: The code has been updated with this approach

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java
Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java]: Fixed

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

New pull of branch and new local build of docs.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java
Comment (aaneja) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java:157]: Can we instead use `assertEquals(actual.toString(true,true), expected.toString(true,true));` ? Would this be a stronger or weaker assertion ?

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java
Comment (aaneja) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java:147]: Can you deserialize into a new empty state object instead of reusing this `state` object ?

**Review by aaneja — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchWithKAggregationFunction.java
Comment (aaneja) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchWithKAggregationFunction.java:121]: Also: Do we need to hardcode these min/max `k` constants ? Should we instead use  `DEFAULT_M` and `MAX_K` from the `KllSketch` class ?

Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchWithKAggregationFunction.java:121]: `DEFAULT_M` isn't exposed publicly, but I will refer to the `MAX_K` argument

**Review by ClarenceThreepwood — DISMISSED** [review]
lgtm. Just nits from my side
Good work on the memory accounting!

File: presto-docs/src/main/sphinx/functions/sketch.rst
Comment (steveburnett) [reply presto-docs/src/main/sphinx/functions/sketch.rst:38]: I noticed new content had been added addressing this, so I pulled the branch and made a local build of the docs just now. Here's a screenshot of the newly-added `sketch_kll_with_k` entry:
Screenshot 2024-01-24 at 9 36 22 AM

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java
Comment (aaneja) [reply presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java:158]: How about `assertEquals(actual.toByteArray(), expected.toByteArray());` ? This works with the current implementation and asserts that the conversion via the new classes is lossless ?

**Review by aaneja — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationFunction.java
Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationFunction.java:33]: I could see a case where it might be desirable to have it be configurable, but that's why we have the `sketch_kll_with_k` function so users can pass any value they want. I put this here to (1) extract constant value from all of the call sites and (2) provide a stable default in the case the library changes the default after an upgrade. A different value will affect the storage size of the sketch, so I wanted something that would be stable within the Presto code base.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchStateSerializer.java
Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchStateSerializer.java:70]: They might not - since `state` is mutable, Line 67 creates a new sketch which may have a different size than the result of `state.getSketch()` at that time. So, we need to subtract the current state (Line 68), set the new state (line 69) and re-calculate the size (Line 70)

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java
Comment (ZacBlanco) [reply presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java:158]: I just tested this, but received failures. The byte arrays are not equivalent, but still produce the same results. I changed the equality check to use the same fields as in the library method instead.

Comment (ZacBlanco) [reply presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestKllSketchStateSerializer.java:158]: Ah I realized I was using the wrong test method beforehand. I checked again with the `toByteArray` methods on this test and it works. I will update it.

**Comment by tdcmeehan** [comment]
I think we need to document the serialized binary format, similar to T-Digest and QDigest.  Otherwise, they're completely unreadable by any other system.

**Comment by ZacBlanco** [comment]
> I think we need to document the serialized binary format, similar to T-Digest and QDigest.

The serialized binary format is specified by the Apache Datasketches library. This is already noted in the docs changes.

https://github.com/prestodb/presto/pull/21568/files#diff-33bc3be2c487c146b213ef7b64ddc7c91956aa6e8b5d6207b217061312286b37R47-R48

If this is confusing I can try to update the phrasing

**Comment by tdcmeehan** [comment]
I can't find where they explicitly document the binary format--is there a link for that?

**Comment by ZacBlanco** [comment]
The binary format is not documented unfortunately. However, all versions (cpp+java) of the datasketches libraries can serialize and deserialize the sketches. e.g. java can deserialize cpp serialized sketches, and vice versa. AFAIK this is the case for all sketches coming from the datasketches library.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:51]: This entire class seems to be missing visibility modifiers. Could we add them?

**Review by tdcmeehan — COMMENTED** [review]
Just one last thing

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java
Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:51]: Because `KllSketchAggregationState` is an interface, modifiers on fields are redundant (`static` by default) and `private/public` are not allowed. Methods and inner classes are also public by default, hence adding would be redundant.

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/sketch/kll/KllSketchAggregationState.java:51]: Oh, missed that!

**Review by tdcmeehan — APPROVED** [review]

**Comment by ZacBlanco** [comment]
Just for some continuity on the discussion of sketch serialized format compatibility, we brought this to their community in hopes to clarify the guarantees that the maintainers provide. This resulted in some changes to the datasketches documentation in https://github.com/apache/datasketches-website/pull/162 and subsequently https://github.com/apache/datasketches-website/pull/163

**Comment by tdcmeehan** [comment]
Possible follow up: if KLL has a provable error bound, could we use it to replace approx_distinct?  The only reason approx_distinct uses Q-Digest under the hood, in spite of its poor performance compared to T-Digest, is it has a provable error bound.

**Comment by ZacBlanco** [comment]
I'm not familiar with the code for `approx_distinct`, but if the KLL sketch answers the same question as the Q-digest (e.g. what percentile does value X fall? or the inverse), then we should be able to replace it. Another option is you could replace the `approx_distinct` implementation entirely with a Theta sketch. I can't speak how the accuracy of a Theta sketch compares to the current `approx_distinct` implementation though.

This could be a good project/experiment for someone new to Presto

**Comment by jmalkin** [comment]
The binary format is defined here: https://github.com/apache/datasketches-java/blob/master/src/main/java/org/apache/datasketches/kll/KllPreambleUtil.java

Obviously the format for items cannot be defined for arbitrary type T, and the exact size of the levels array is not fixed. But the serialized header is defined, as is the order of the non-fixed-size/format data.

**Comment by leerho** [comment]
@Yuhta,
I don't understand your comment above (Jan 4th) "The serialized form is not very compact and can occupy more memory than it needs."

The KLL sketch always serializes (`toByteArray()`) into a compact form, which has no wasted space,  and grows very gradually based on the number of items fed to and retained by the sketch.  You can see the effect of this in the plot copied from our website above.  The blue curve starts very small (8 bytes) and grows gradually until it reaches ~2.5K bytes at about 1M items.

It is true that if the sketch image were compressed it would become even smaller, but that can be costly in terms of time.  That is not an option we currently offer as part of the sketch, but as long as one is using lossless compression algorithms, it is something you could do yourself. 

If you could elaborate you concern a bit more it would help us understand the issue.

Thanks!

**Pull request merged**
